Human readable dumps of formulas

facundominguez opened the pull request:
This PR addresses most of ucsd-progsys/liquidhaskell#1848

With `--savequery` liquid haskell now saves files .liquid/module_name.fq.prettified

The format is not valid .fq. Instead, it shows a list of constraints in the form:

```
constraint:
  lhs {VV##F##7 : Tuple | [(~ ((isAdmit Language.Haskell.Liquid.ProofCombinators.QED)))]}
  rhs {VV##F##7 : Tuple | [(ty##aB0 = (Type.TFun (Type.funArgTy ty##aB0) (Type.funResTy ty##aB0)))]}
  id 7 tag [6]
  // META constraint id 7 : Type.hs:(41,3)-(45,5)
  environment:
    // elided some likely irrelevant bindings
    
    ty##aB0 : {VV##0 : Type.Ty | [(Type.isFunTy VV##0)]}
    
  ...
```

The environment of each constraint is populated from the FInfo at the time the query is saved.
The prettified output differs from the .fq output in the following ways.

### 1. Bindings prefixed with `lq_anf` are sometimes inlined at their use sites

Given a binding like `a : { v | e1 }`, expressions of the form `... e0 = a ...` are changed to `... e1[a:=e0] ...`

`... e0 = (a : s) ...` is equally transformed to `... e1[a:=e0] ...` and the sort annotation is dropped.

Given a binding like `a : { v | v = e1 }`, expressions of the form `... a ...` are changed to `... e1 ...` if `v` does not appear free in `e1`.

### 2. Refinement types of boolean sort are simplified

Refinement types of the form `{v:bool | v && P v}` become `{v:Bool | v && P true}`.

Refinement types of the form `{v:bool | ~v && P v}` become `{v:bool | ~v && P false}`.

### 3. Bindings considered unrelated to the constraint are dropped

Say that we have an environment like
```
a : { v | v > b }, b : int, c : { v | v >= b && b >= d}, d : int, e: { v | P }
```
and a constraint
```
{ true } <: {a > 0}
```

We retain `a` because it is used in the constraint. 
We retain `b` because it is used by the binding of `a`. 
We retain `c` because it states facts about `b`.
We retain `d` because it is used in `c` which is also retained.

The only binding that we drop is `e` as long as `P` doesn't mention any retained binding.
Dropping `e` can still change the outcome of verification (suppose that `P` is `false`). However, we drop the binding anyway because this output is to be consumed by a human, not a computer. And humans will often inspect this output when `constraints` *fail* to verify, which means that the environment can't be inconsistent as necessary to invalidate this binding elimination.

Nonetheless, because of this transformation, we warn the user that some bindings are elided.

### 4. Unused binding names are erased

Bindings of the form `a : {v | P }` become `_ : { v | P }` if `a` is not used in any binding in the environment nor in the constraint.

### 5. Some extra simplifications of predicates

`P => true` simplifies to `P`
`P => false` simplifies to `~P`
`~~P` simplifies to `P`
`true <=> P` simplifies to `P`
`false <=> P` simplifies to `~P`
`P && P` simplifies to `P`

facundominguez committed:
Implement environment reduction of FInfos

Also produces more readable formulas in .fq.prettified files

facundominguez committed:
Expose exprSymbolsSet

facundominguez committed:
Simplify PIff

facundominguez committed:
Simplify double negation and implication

ranjitjhala commented:
wow this is AWESOME thank you!!! (waiting for the tests to pass...)

facundominguez commented:
In my tests, these transformations reduce the environments to a tenth of what they were.

I still expect to iterate a few times on this code as I try it on more examples. I'm fine doing it in this PR or on following ones.

I'm opening the PR to ask for advice on how to best integrate it with the rest of liquid-fixpoint. Maybe some functions are better called from elsewhere, or maybe I'm duplicating some existing transformations, and there is also a function I'm using to combine the predicates of bindings that have the same name appearing in an environment. See the function [mergeDuplicateBindings](https://github.com/ucsd-progsys/liquid-fixpoint/pull/473/files#diff-8dedee7c76f413897144d7b36fd39caee492cc4f17e2ec211f2845fbf865b792R118-R124)

```
-- | If the environment contains duplicated bindings, they are
-- combined with conjunctions.
--
-- This means that @[ a : {v | P v }, a : {z | Q z }, b : {v | S v} ]@
-- is combined into @[ a : {v | P v && Q v }, b : {v | S v} ]@
--
mergeDuplicatedBindings :: [(Symbol, SortedReft)] -> HashMap Symbol SortedReft
```

Also, I'm changing the `simplify` method of `Fixpoint Expr` which can affect other code. So far, this method looks rather conservative on what it tries to simplify.

ranjitjhala commented:
Hi @facundominguez -- this is super! To integrate, I suggest you try to squeeze this simplification step into one of the steps in `Sanitize` shown here

https://github.com/ucsd-progsys/liquid-fixpoint/blob/develop/src/Language/Fixpoint/Solver/Sanitize.hs#L42

this is a bunch of pre-processing transforms done *before* solving. 

There are of course, various (unspecified :-( ) invariants about the `SInfo` structure -- but once you insert your `prettify` into the pipeline we can use the fixpoint and LH battery of tests to make sure nothing strange is happening. It would be a *huge* win if we could greatly shrink the environments!

(Does that make sense?)

facundominguez commented:
I've numbered the transformations in the description.

I think these transformations are safe to put in the verification pipeline:
* (1) Bindings prefixed with lq_anf are sometimes inlined at their use sites
* (2) Refinement types of boolean sort are simplified
* (3) Bindings considered unrelated to the constraint are dropped (safe version)
* (5) Some extra simplifications of predicates

Since these transformations act on the environments of constraints, each constraint would need to go through them separately. 
I would insert these transformations at the beginning of `sanitize`, and then the prettified output could dump the constraints either immediately or at the end of sanitize.

The version of (3) intended for human consumption would remove just too much. But we could use a more conservative version implemented in the function [_dropIrrelevantBindings](https://github.com/ucsd-progsys/liquid-fixpoint/pull/473/files#diff-8dedee7c76f413897144d7b36fd39caee492cc4f17e2ec211f2845fbf865b792R223-R232).

That said, what is the purpose of reducing the environment at this stage? Despite having large environments, the SMT solver is much faster than liquid-haskell+liquid-fixpoint. Aren't we making `liquid-fixpoint` even slower by enlarging the verification pipeline? Maybe sanitize is very expensive (?), I don't have any profiling numbers this far.

ranjitjhala commented:
> That said, what is the purpose of reducing the environment at this stage? Despite having large environments, the SMT solver is much faster than liquid-haskell+liquid-fixpoint

Great question -- `liquid-fixpoint` makes lots of traversals over the environments to generate the SMT queries -- so if we can pay up front to drastically shrink the environments, it could save a lot in the cost of the downstream traversals. (Of course, we won't know for sure until it's actually implemented...)

Does that make sense?

facundominguez commented:
Alright, let's try it. I'm writing the glue code.

facundominguez commented:
@ranjitjhala, do you have any clue of what could cause such an innocent change as c7a7f37 to stop `Tests.Unit.proof.pleBoolC.fq` from blocking in CI?

ranjitjhala commented:
That is very strange -- can you post the "before" and "after" .fq files for those two? Maybe that will give a clue?

facundominguez commented:
When writing the glue code two questions arose.

1. `GInfo` has a field `ebinds :: [BindId]`. How should this field change when the referenced bindings are removed or changed in the global environment?
2. `GInfo` also has a field `bindInfo :: HashMap BindId a`. Similar question to above. Should metadata just be dropped if bindings are changed or removed?

ranjitjhala commented:
hmm yes -- in general, those need to be around to determine the `Sort` of the binder -- but if the binder itself has been eliminated, then you can get rid of those too. (As long as they're _really_ eliminated or else there will be odd out of scope errors...)

facundominguez commented:
> but if the binder itself has been eliminated then you can get rid of those too

Can I get rid of these references if the binding was eliminated but also inlined in another binding?

Should I preserve them if the binding is preserved but I applied some substitution or simplification on it?

ranjitjhala commented:
Think of it as a "free variable" so if the variable still "occurs" elsewhere, you want to leave it.

But if there are "no occurrences" then you can remove it.

ranjitjhala commented:
Maybe a concrete example will help?

On Fri, Jul 16, 2021 at 10:18 AM Facundo Domínguez ***@***.***>
wrote:

> but if the binder itself has been eliminated then you can get rid of those
> too
>
> Can I get rid of these references if the binding was eliminated but also
> inlined in another binding?
>
> Should I preserve them if the binding is preserved but I applied some
> substitution or simplification on it?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D881599180&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=BOexYp5xcoSH7ySXItR0C7P-aCwlnlydLo39ZTLccwE&s=mpy8zPr_lk0k909n55Qpp4ugI_TIFIF9D3gj_iAAinA&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OHQC45QZ2HEKAYBNLDTYBSW5ANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=BOexYp5xcoSH7ySXItR0C7P-aCwlnlydLo39ZTLccwE&s=i7fflOmaJEIMwNefg9itAX6SZC9S3vYW4B-MyxKrFpg&e=>
> .
>

facundominguez commented:
Thanks Ranjit!

I admit I still have to figure out the precise meaning of `ebinds` and `bindInfo` to handle them with confidence. But let me return to you when I inspect the code some more.

On the failing test, I found that the simplification that makes the pleBoolC.fq block is equivalent to the following transformation of the input:
```
diff --git a/tests/proof/pleBoolC.fq b/tests/proof/pleBoolC.fq
index 2efb3cfc..5161d4e9 100644
--- a/tests/proof/pleBoolC.fq
+++ b/tests/proof/pleBoolC.fq
@@ -27,8 +27,8 @@ constant tail: (func(1, [Vec @(0); Vec @(0)]))
 
 
 define mall(f:func(0,[int;bool]), xs:Vec int) : bool = {
-    if (isNil xs) then true 
-    else (mand (f (head xs)) (mall f (tail xs)))
+    (~(isNil xs)) => (mand (f (head xs)) (mall f (tail xs)))
 }
 
 define mand(x:bool, y:bool) : bool = { 
```
The above blocks in the `develop` branch when invoked as 
```
$ fixpoint --eliminate=some liquid-fixpoint/tests/proof/pleBoolC.fq
```
Is this a bug?

ranjitjhala commented:
haha I would call it more of a "missing feature" :-) 

The PLE algorithm _should_ be able to deal with the boolean-ized version, but right now, still uses the `if-then-else` syntactic structure. 

So, for now I'd suggest either

* Don't modify if-then-else structure (so simplify under the branches, but leave the `if-then-else` in place) OR
* Don't modify the _function definitions_ i.e. the `define` stuff.

facundominguez committed:
Don't simplify implication to avoid PLE divergence

facundominguez commented:
The desugaring of `if-then-else` could be tricky to identify in expressions, so I'm just refraining from simplifying implication. The alternative to turn off simplifications in function definitions looked too extreme for our immediate needs.

facundominguez committed:
Implement fromListBindEnv

facundominguez commented:
I placed most of the simplifications in a function `reduceEnvironment` which is called before `sanitize`. The prettified output is printed after `reduceEnvironment` and before `sanitize`.

Now, there are a few failing tests. The first of them is `tests/pos/ebind-01.fq` which I don't know why it fails. The output of `reduceEnvironment` looks sensible to me, yet the verbose output of `fixpoint` differs. I paid particular attention to the binding environment and the `ebinds` field.

Without `reduceEnvironment` it reads:
```
$ fixpoint liquid-fixpoint/tests/pos/ebind-01.fq -v
...
**** DONE:  Worklist Initialize ************************************************
 
Working   0% [.................................................................]
Computing Result
SMT READ 25% [=================
SMT Says: Unsat
UNSAT id 1 False
LHS: false
RHS: VV##1 == x1
SMT READ 50                    ================
SMT Says: Unsat
UNSAT id 2 False
LHS: exists [VV##1 : int]
       . VV##1 == m + 1
         && VV##1 == x1
     && VV##2 == x1 + 1
RHS: VV##2 == 2 + m
RESULT: 1afe (Stats {numCstr = 2, numIter = 2, =================================================================]
Saving Solution: liquid-fixpoint/tests/pos/.liquid/ebind-01.fq.fqout

Safe ( 2  constraints checked)
```

with `reduceEnvironment` it says

```
$ fixpoint liquid-fixpoint/tests/pos/ebind-01.fq -v
...
**** DONE:  Worklist Initialize ************************************************

Working   0% [.................................................................]
Computing Result
SMT READ 25% [=================
SMT Says: Unsat
UNSAT id 1 False
LHS: false
RHS: VV##1 == x1
SMT READ 50                    ================
SMT Says: Sat
UNSAT id 2 True
LHS: exists [m : int]
       . VV##1 == m + 1
         && VV##1 == x1
     && VV##2 == x1 + 1
RHS: VV##2 == 2 + m
RESULT: 1nsafe (Stats {numCstr = 2, numIter = 2=================================================================]
Saving Solution: liquid-fixpoint/tests/pos/.liquid/ebind-01.fq.fqout

Unsafe:
WARNING: 2
```

Any hints of what to look for here?

ranjitjhala commented:
Super! The deal with the `ebind` is that while regular bind are "universal", these are "existential". 

That is under the usual "universal", the meaning of the .fq is the logical formula

```
forall m, 
  forall x1. 
     forall v. v = m + 1 => v = x1
  /\ forall v. v = x1 + 1 => v = 2 + m
```

which is clearly *invalid*. However, under the "existential" interpretation, the .fq means


```
forall m, 
  exists x1. 
     forall v. v = m + 1 => v = x1
  /\ forall v. v = x1 + 1 => v = 2 + m
```
  
which _is_ valid (as `x1` is the `m + 1`).

So, my question is when you say

>  The output of reduceEnvironment looks sensible to me

Can you paste in what exactly that output is?

facundominguez commented:
> Can you paste in what exactly that output is?

There you have
```
expand []
rewrite  []

bind 0 VV##1 : {VV##1 : int | [(VV##1 = (m + 1))]}
bind 2 m : {v : int | []}
bind 3 VV##2 : {VV##2 : int | [(VV##2 = (x1 + 1))]}

ebind 1 x1 : {  int  }

simpleConstraint:
  env [0; 1; 2]
  rhs (VV##1 = x1)
  id 1 tag []
  // META simpleConstraint id 1 : ()

simpleConstraint:
  env [1; 2; 3]
  rhs (VV##2 = (2 + m))
  id 2 tag []
  // META simpleConstraint id 2 : ()
```

facundominguez commented:
Alas, flipping the bind ids breaks the test in the same way if I don't reduce the environment.

```
diff --git a/tests/pos/ebind-01.fq b/tests/pos/ebind-01.fq
index 65baed64..ca6057dd 100644
--- a/tests/pos/ebind-01.fq
+++ b/tests/pos/ebind-01.fq
@@ -1,7 +1,7 @@
 fixpoint "--eliminate=all"
 
-bind  1 m  : {v: int | true }
-ebind 2 x1 : { int }
+bind  2 m  : {v: int | true }
+ebind 1 x1 : { int }
 // bind  1 x1 : {v: int | v = m + 1 }
 
 constraint:

```
Are ebind ids supposed to be higher numbered than regular bindings? Can that be relaxed?

ranjitjhala commented:
Ah no the order matters

Because it’s for all-exists vs exists-for all which have different
semantics.

However - the ebind thing is a highly experimental feature that is not used
by ANY of the LH benchmarks.

Maybe it is easiest to disable your simplifications if there are any ebind?



On Wed, Jul 21, 2021 at 1:01 PM Facundo Domínguez ***@***.***>
wrote:

> Alas, flipping the bind ids breaks the test in the same way if I don't
> reduce the environment.
>
> diff --git a/tests/pos/ebind-01.fq b/tests/pos/ebind-01.fq
> index 65baed64..ca6057dd 100644
> --- a/tests/pos/ebind-01.fq
> +++ b/tests/pos/ebind-01.fq
> @@ -1,7 +1,7 @@
>  fixpoint "--eliminate=all"
>
> -bind  1 m  : {v: int | true }
> -ebind 2 x1 : { int }
> +bind  2 m  : {v: int | true }
> +ebind 1 x1 : { int }
>  // bind  1 x1 : {v: int | v = m + 1 }
>
>  constraint:
>
>
> Are bind ids supposed to be sorted? Can that be relaxed?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D884460574&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=PnuR4dqfnEdutf90MH3yxbeLVNHZT0IWThLyJAZ6pso&s=cYYdMuCnDw68NwOAiqxEF1ReVAC_JBLwq9QIN39zDMQ&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OGXWRDIFCGDZJITL73TY4RQ7ANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=PnuR4dqfnEdutf90MH3yxbeLVNHZT0IWThLyJAZ6pso&s=wgn4iWmHOECs-V8F1jK8IluN5T_OjB46Ywh9n9N15js&e=>
> .
>

facundominguez commented:
If the numbering determines the order in which quantifiers are introduced, I cannot test it reliably. I always get constraints rejected when starting with a constraint with an existential.

I'm willing to disable simplifications when existentials are present, until I get a better understanding of how the feature works.

I'm thinking of printing a warning message whenever simplifications are skipped (and honor -v0), unless you consider that too much noise.

ranjitjhala commented:
That sounds great to me!

facundominguez committed:
Implement onEverySubexpr

facundominguez committed:
Move exprSymbolsSet to Refinements.hs

facundominguez committed:
Implement sortSymbols and sortedReftSymbols

facundominguez commented:
I just discovered that I can't use the same strategy to discover relevant bindings for well-founded constraints and subtyping constraints. Consider for instance `tests/pos/bool03.fq`

```
qualif LE(v:a, x:a): (x <= v)

constant lit$36$not$45$the$45$hippopotamus : (Str)
constant lit#cat : (Str)
constant lit#dog : (Str)

bind 1 bx : {v: bool | true }

bind 2 zx : {v: bool | lit#cat /= lit#dog }

constraint:
  env [ 1 ]
  lhs {v : bool | bx <= v }
  rhs {v : bool | $k1     }
  id 1 tag []

constraint:
  env [ 1; 2 ]
  lhs {v : bool | $k1 }
  rhs {v : bool | bx <= v }
  id 2 tag []

wf:
  env [1]
  reft {v : bool | $k1 }
```
In the `wf` constraint, the rhs doesn't mention any bindings in the environment `[1]`. Yet, stripping the environment would prevent the liquid variable from being instantiated to the necessary solution (`bx <= v`). I'm going to try to drop bindings only when they don't show up in any other constraint where the liquid variable is used. Also, if two liquid variables are used in a single constraint, they both could depend on the bindings of each other.

facundominguez committed:
Implement exprKVars

facundominguez commented:
> I'm going to try to drop bindings only when they don't show up in any other constraint where the liquid variable is used. Also, if two liquid variables are used in a single constraint, they both could depend on the bindings of each other.

Works on a few cases, but it fails when substitutions are involved like in `qualif-template-02.fq`.
```
...
bind 0  sunday  : {v : int | v = 6  }
bind 1  monday  : {v : int | v = 4  }
bind 2  tuesday : {v : int | v = 10 }

bind 10 x       : {v : int | v = 4  }
bind 11 y       : {v : int | v = 6  }
bind 12 z       : {v : int | v = 10 }
...
constraint: 
  env [10; 11; 12]
  lhs {v : int | $k0[sunday := x][monday := y][ tuesday := z] }
  rhs {v : int | v = 10 }
  id 2 tag []

wf:
  env [0; 1; 2]
  reft {v : int | $k0}
```
See how  `k0` is used in the constraint 2. However, none of the bindings 0, 1, and 2 of `wf` are used in constraint 2, and therefore they are considered irrelevant. Looks like I need to pay attention to the symbols used in the substitutions as well.

facundominguez commented:
Past the last error, the next failing test is `tests/proof/ple1.fq`
```
fixpoint "--rewrite"

constant foo: (func(1, [@(0)  ; int]))
constant bar: (func(0, [Bob   ; int]))

define foo(x : alpha) : int = { bar (coerce (alpha ~ Bob) x) }
define bar(y : Bob)   : int = { 22 } 

expand [1 : True]

bind 0 z : {v: beta | true }

constraint:
  env []
  lhs {v : int | true }
  rhs {v : int | (foo z) = 22 }
  id 1 tag []
```
@ranjitjhala, how is that `z` doesn't need to be in the environment of constraint (1) in order to be used?

ranjitjhala commented:
!!!! Wow that’s a real surprise to me, it definitely needs to be there!

On Sat, Jul 24, 2021 at 12:34 PM Facundo Domínguez ***@***.***>
wrote:

> Past the last error, the next failing test is tests/proof/ple1.fq
>
> fixpoint "--rewrite"
>
> constant foo: (func(1, [@(0)  ; int]))
> constant bar: (func(0, [Bob   ; int]))
>
> define foo(x : alpha) : int = { bar (coerce (alpha ~ Bob) x) }
> define bar(y : Bob)   : int = { 22 }
>
> expand [1 : True]
>
> bind 0 z : {v: beta | true }
>
> constraint:
>   env []
>   lhs {v : int | true }
>   rhs {v : int | (foo z) = 22 }
>   id 1 tag []
>
> @ranjitjhala
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ranjitjhala&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=bJAuIlN8xcGPrLDAC5mQpI6nI3UhPyeJ7FOdFW3D2k8&s=nJmu0NRfSYUfCX5VQrKlbMEt5uc16alGOERPPpCu9DE&e=>,
> how is that z doesn't need to be in the environment of constraint (1) in
> order to be used?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D886101916&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=bJAuIlN8xcGPrLDAC5mQpI6nI3UhPyeJ7FOdFW3D2k8&s=b9mLxzcU8zpSbVixoWa6CKfV1Eg-He6rNIwBw_qLux4&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OFW2LNXCK45LJKCIBLTZMIT5ANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=bJAuIlN8xcGPrLDAC5mQpI6nI3UhPyeJ7FOdFW3D2k8&s=xzEaVkpC3gFHe0WJkuJZFc5zRICxoAnHCWPKBtimer8&e=>
> .
>

facundominguez commented:
In this case the simplification breaks the test because it removes the binding `z` (since it isn't referred from any constraint environment). Perhaps I could arrange for the binding to be retained when the constraint refers to `z` and it is missing from the environment, but it would be simpler if we just required the binding to be present in the constraint.

ranjitjhala commented:
Hi @facundominguez - I think it’s simpler to just fix the test by adding 0 to the env which is the correct thing to do here.

facundominguez committed:
Fix environment of ple1.fq test

facundominguez commented:
The last failing test in the test-suite is `tests/proof/rewrite.fq`. Constraint (3) is rejected as unsafe after simplification.
```
...

bind 0 xs    : {v: List | true }
bind 1 ys    : {v: List | true }
bind 2 zs    : {v: List | true }
bind 3 ws    : {v: List | true }

constraint:
  env [0; 1; 2; 3]
  lhs {v1 : List | true  }
  rhs {v2 : List | concat3Left (concat xs ws) ys zs = concat3Right (concat xs ws) ys zs }
  id 1 tag []

constraint:
  env [0; 1; 2; 3]
  lhs {v1 : List | true  }
  rhs {v2 : List | concat3Right (concat xs ws) ys zs = concat3Left (concat xs ws) ys zs }
  id 2 tag []

constraint:
  env [0; 1; 2; 3]
  lhs {v1 : List | true  }
  rhs {v2 : List | concat3Left (concat xs ws) ys zs = concat3Right xs ws (concat ys zs) }
  id 3 tag []
```
The simplification is not removing anything, the only thing that it does is shuffling the identifiers of the bindings.

Moreover, skipping the simplification and removing constraint (2), also has constraint (3) rejected in a similar way (update: I couldn't reproduce this later, probably I was doing something wrong). I find the failure a bit puzzling, and I was wondering if this could be a bug in ple? Anything else I should look for?

Here's the simplified output: [rewrite.fq.fq.sinfo.txt](https://github.com/ucsd-progsys/liquid-fixpoint/files/6874681/rewrite.fq.fq.sinfo.txt)
And here's the non-simplified output: [rewrite.fq.fq.sinfo.txt](https://github.com/ucsd-progsys/liquid-fixpoint/files/6874686/rewrite.fq.fq.sinfo.txt)

ranjitjhala commented:
Hmm not sure what’s happening here @nikivazou or @zgrannan can you help?

ranjitjhala commented:
@facundominguez - another option is to disable the simplification when we have those auto rewrite binds?

zgrannan commented:
@facundominguez This test is for an (undocumented, I believe) extension of PLE that enables the equality of an  `autorewrite` to be automatically inserted into the SMT environment when PLE evaluates an expression that matches the left-hand-side.

In this test for example, the third constraint should  be resolved by instantiation of the 1st autorewrite, since rewriting should generate the equalities:
```
    concat (concat (concat xs ws) ys) zs
=== concat (concat xs ws) (concat ys zs)
=== concat xs (concat ws (concat ys zs))
```

I'm not sure where the actual bug is in this case. Recently I discovered some bugs recently in the "autorewrite" functionality, but I won't have a PR to address them at least for a few weeks since I'm about to go on vacation.

As @ranjitjhala suggested removing the simplification in the presence of "autorewrite" would be a reasonable workaround I think.

facundominguez commented:
Thanks @zgrannan and @ranjitjhala!

I'll try disabling simplification in this case too.

The test seems to be sensitive to the order of the BindIds after further experimentation. The test passes if all the lhss are given the highest BindIds. All three constraints are rejected if the lhss are given the lowest BindIds.

facundominguez committed:
Implement a flag to disable environment reduction.

facundominguez committed:
Implement dedupAutoRewrites

facundominguez commented:
The liquid-fixpoint tests pass, but I'm finding some modules in stitch-lh that reject some constraints when environments are reduced. Also I still have to try this on the testsuite of liquidhaskell.

Verification times are little changed by the environment reductions in the modules I've tried from stitch-lh. Perhaps you can come up with better examples.

On the other hand, the prettified output is glorious every time I checked. Hundreds of bindings get reduced to 10 or 20.

facundominguez commented:
Several tests fail in LH, so it is going to be some time until I fix them all. But on the bright side, the benchmarks do show noticeable improvements. Here there are some. The times in parenthesis correspond to LH unmodified, and times at the right correspond to LH + environment reduction.

```
    esop
      Base.hs:                                OK (68.81s) vs 56s
    vect-algs
      Data/Vector/Algorithms/Merge.hs:        OK (22.72s) vs 16s
    bytestring
      Data/ByteString/Fusion.T.hs:            OK (66.12s) vs 15s
      Data/ByteString/Lazy.hs:                OK (59.65s) vs 28s
      Data/ByteString/Lazy/Char8.hs:          OK (12.70s) vs 8s
      Data/ByteString/Char8.hs:               OK (29.04s) vs 17s
    text
      Data/Text/Search.hs:                    OK (56.54s) vs 17
      Data/Text/Fusion.hs:                    OK (23.36s) vs 13
```
I didn't spot any benchmark worsening as dramatically just yet. So I guess there is no escaping to continue fixing the failing tests :sob: 


@ranjitjhala is the complexity introduced so far bearable?

ranjitjhala commented:
Holy smokes -- that's amazing!!!

ranjitjhala commented:
This in particular is super amazing!

```
      Data/ByteString/Fusion.T.hs:            OK (66.12s) vs 15s
```

ranjitjhala commented:
A quick warning, btw.

I noticed that due to various reasons (module loading/caching of results from imported modules) you can get quite different times if you "re-run" the benchmark, because the "first" run analyzes the imports, and the "second" run can skip doing so :-)

So to double check just wipe the `.liquid` folders e.g.

```
$ cd benchmarks/text-.../
$ rm -rf **/.liquid
```

between the two runs...

facundominguez commented:
Thanks. That's good to know. I've checked that the times reported above are still reproducible when preceding the benchmarks with
```bash
$ find benchmarks \( -name "*.hi" -o -name "*.o" -o -name .liquid \) -exec rm -r {} ";"
```

ranjitjhala commented:
Outstanding! (These numbers have put me in a _very_ good mood this morning, thank you!!!!)

facundominguez commented:
I discover one more bug in environment reduction today.

When dropping bindings, I haven't been paying attention to `define` clauses. So if a constraint refers to a `define`d identifier, and the definition refers to a binding in the environment, I might just drop the binding anyway.

I presume that this also could happen with `match`es.

ranjitjhala commented:
I don’t think I follow entirely. The define and match clauses are always
self-contained. That is they have no free variables. Do you have an example
of the kind of thing you’re talking about?

On Thu, Jul 29, 2021 at 2:01 PM Facundo Domínguez ***@***.***>
wrote:

> I discover one more bug in environment reduction today.
>
> When dropping bindings, I haven't been paying attention to define
> clauses. So if a constraint refers to a defined identifier, and the
> definition refers to a binding in the environment, I might just drop the
> binding anyway.
>
> I presume that this also could happen with matches.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D889454244&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=8_OQCX-alzMrAIuiTlayVQqb25yHcuRQIJ1Uptay4Bs&s=MtJ-nwQRLQCP4rbAhoP8Cgu__b5GwGfVl-BOKXP0XDc&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OCULTS36Q7S5LGERNTT2G6Q7ANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=8_OQCX-alzMrAIuiTlayVQqb25yHcuRQIJ1Uptay4Bs&s=jshydy0J5LKhezS-kbDIX-Pid5BVZvcnERGLg-ra9Bw&e=>
> .
>

facundominguez commented:
Verification of [this file](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/tests/import/lib/RewriteLib.hs) fails when environment reduction is enabled. I think this is because I'm dropping the bindings of `[]` and `(:)`, the list data constructors. They are not referenced directly in the constraint or other bindings, but they are used in the `define` clause of `(++)`.

ranjitjhala commented:
👍🏽

On Thu, Jul 29, 2021 at 3:01 PM Facundo Domínguez ***@***.***>
wrote:

> Verification of this file
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquidhaskell_blob_develop_tests_import_lib_RewriteLib.hs&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=n7wdLa0yuAuEupk7e1xu9XEqoJPYHx1Q6Nzsa0VNh-g&s=Ap4wtXnkfevUjfrkvpQ7SVMhLrgblrZBc5oFIO5u2do&e=>
> fails when environment reduction is enabled. I think this is because I'm
> dropping the bindings of [] and (:), the list data constructors. They are
> not referenced in formulas, but they are used in the define clause of (++)
> .
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D889488951&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=n7wdLa0yuAuEupk7e1xu9XEqoJPYHx1Q6Nzsa0VNh-g&s=imjXEPQNoeSzGPrOkKbJYaucJz59bb4oBUqmNNoyyFc&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OCKQCAIXVU37JL3INLT2HFVJANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=n7wdLa0yuAuEupk7e1xu9XEqoJPYHx1Q6Nzsa0VNh-g&s=vBOZkVT9-MtKvh5cKjKQ1wAOqxaJ8ToSD0meiucAQuw&e=>
> .
>

facundominguez commented:
Hello! I found some more tests where the binding numbering matters (similarly to what we found in https://github.com/ucsd-progsys/liquid-fixpoint/pull/473#issuecomment-886271459). I reduced one of the tests:

```
data ReflectClient4a.List 1 = [
       | ReflectClient4a.Cons {ReflectClient4a.Cons##lqdc##$select##ReflectClient4a.Cons##1 : @(0), ReflectClient4a.Cons##lqdc##$select##ReflectClient4a.Cons##2 : (ReflectClient4a.List @(0))}
       | ReflectClient4a.Nil {}
     ]

match ReflectClient4a.llen ReflectClient4a.Nil  =  (0)
match ReflectClient4a.llen ReflectClient4a.Cons lq_tmp$x##377 lq_tmp$x##378  =  ((1 + (ReflectClient4a.llen lq_tmp$x##378)))
expand [13 : True; 14 : True]

constant ReflectClient4a.Cons : (func(1 , [@(0);
                                           (ReflectClient4a.List @(0));
                                           (ReflectClient4a.List @(0))]))
constant ReflectClient4a.Nil : (func(1 , [(ReflectClient4a.List @(0))]))
constant ReflectClient4a.llen : (func(1 , [(ReflectClient4a.List @(0)); int]))

distinct ReflectClient4a.Cons : (func(1 , [@(0);
                                           (ReflectClient4a.List @(0));
                                           (ReflectClient4a.List @(0))]))
distinct ReflectClient4a.Nil : (func(1 , [(ReflectClient4a.List @(0))]))


bind 0 ReflectClient4a.llen : {VV : func(1 , [(ReflectClient4a.List @(0)); int]) | []}
bind 1 GHC.Tuple.$40$$41$ : {VV##404 : Tuple | [(VV##404 = GHC.Tuple.$40$$41$)]}
bind 4 ReflectClient4a.Cons : {VV : func(1 , [@(0); (ReflectClient4a.List @(0)); (ReflectClient4a.List @(0))]) | []}
bind 5 ReflectClient4a.Nil : {VV : func(1 , [(ReflectClient4a.List @(0))]) | []}

constraint:
  env [0; 1; 4; 5]
  lhs {VV##13 : Tuple | [(VV##13 = GHC.Tuple.$40$$41$)]}
  rhs {VV##13 : Tuple | [((ReflectClient4a.llen (ReflectClient4a.Cons 2 ReflectClient4a.Nil)) = 1)]}
  id 13 tag [7]
  // META constraint id 13 : ()

constraint:
  env [0; 1; 4; 5]
  lhs {VV##14 : Tuple | [(VV##14 = GHC.Tuple.$40$$41$)]}
  rhs {VV##14 : Tuple | [((ReflectClient4a.llen ReflectClient4a.Nil) = 0)]}
  id 14 tag [8]
  // META constraint id 14 : ()
```

When this example is run we get
```
$ fixpoint ReflectClient4a.hs.fq --eliminate=some --rewriteaxioms --nonlincuts --allowho
...
Unsafe:
WARNING: 13
WARNING: 14
```
But the constraints are accepted if:
1) I disable environment reduction, or
2) I remove one of the constraints from the input, or
3) I reorder the bindings so the lhs bindings appear with higher numbers than the preexisting bindings after environment reduction

I couldn't find a way to expose this "bug" without environment reduction, since the lhs bindings are introduced when converting `FInfo ()` to `SInfo ()`, and by default these bindings are higher-numbered.

Environment reduction then rehashes the identifiers and we can see the failures.

I'm going with (3) for now, but ideally someone would figure out why the order of the numbering of the bindings matters.

facundominguez commented:
I created [a PR](https://github.com/ucsd-progsys/liquid-fixpoint/pull/479) with some spin-off changes from debugging this.

facundominguez commented:
And [here](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/tests/classes/pos/RealProps1.hs) I found a test that leaves me head-scratching. There are two sorts assigned to a symbol in the same environment. Some symbol `a` is given sort `num` in one binding and sort `frac` in another.

From the fq file:
```
bind 104 a##a1Am : {VV : frac | []}

bind 106 a##a1Am : {VV : num | []}

...
constraint:
...
       104;
       105;
       106;
       107;
       108;
       109;
       110]
  lhs {VV##F##16 : bool | [(VV##F##16 <=> (lq_anf$##7205759403792803203##d2uT = lq_anf$##7205759403792803204##d2uU))]}
  rhs {VV##F##16 : bool | [VV##F##16]}
  id 16 tag [4]
  // META constraint id 16 : tests/classes/pos/RealProps1.hs:14:1-30
```

Are symbols supposed to have multiple sorts? It would make it challenging to determine which binding is meant when it shows up in a predicate.

ranjitjhala commented:
This particular quirk is because of an old way to hack numeric type classes into fixpoint. 

The binders themselves (a1##a1Am) are special in that they correspond to sorts that appear elsewhere in the environment… let me see if I can dig up the code for this. 

Are you getting a test failure after deleting one or another such bind? There’s probably some way to tell if the binder is actually a sort variable and if so, to leave it be…

facundominguez commented:
> Are you getting a test failure after deleting one or another such bind?

The failure I get is from the code I introduced to consolidate binders with the same name into a single binder. This is only possible if all of the binders have the same sort. Dropping the `frac` bind has the constraints accepted. Dropping the `num` bind produces an error instead:
```
Crash!: :1:1-1:1: Error
  elaborate solver elabBE 35 "VV##16" {VV##16 : bool | [(VV##16 <=> ((x##a1yc * y##a1yd) = (y##a1yd * x##a1yc)))]} failed on:
      VV##16 <=> x##a1yc * y##a1yd == y##a1yd * x##a1yc
  with error
      The sort a##a1Am is not numeric
  in environment
      VV##16 := bool

      x##a1yc := a##a1Am

      y##a1yd := a##a1Am
```

> There’s probably some way to tell if the binder is actually a sort variable and if so, to leave it be…

Ok, that could help.

facundominguez committed:
Derive Show instance for IBindEnv

facundominguez committed:
Implement unionsIBindEnv

facundominguez commented:
Hello! I'm deep into testing environment reduction with LH tests.

There are some tests that break with PIff simplification as implemented in 3d92975d.
```
Tests.Macro.unit-neg.MeasureContains.hs
Tests.Macro.unit-neg.ExactGADT7.hs
Tests.Macro.unit-neg.Ast.hs
```
These programs are supposed to fail verification, yet, PIff simplification makes verification succeed.

`Tests.Macro.unit-neg.MeasureContains.hs` is the smaller of the three. Why is that this test should fail to verify?

Comparing the neg and pos versions I have
```
--- tests/neg/MeasureContains.hs	2021-07-27 10:51:31.203414691 -0300
+++ tests/pos/MeasureContains.hs	2021-08-02 15:52:49.148030497 -0300
@@ -23,4 +21,5 @@
 containsV _               = False
 
 
-prop1 = liquidAssert (containsV $ Other)
+prop1 = liquidAssert (containsV $ V 7)
+prop2 = liquidAssert (containsV $ Bind (M (V 5)) Other)
```
So the pos version actually requires more to verify!

ranjitjhala commented:
Hi @facundominguez  -- it seems that `containsV t` predicate evaluates to `True` if there is _some_ `V i` inside the `t`. Note that the actual assertions in `prop1` are different

*  in the `pos` the assertion is `True` as the term is of the form `V 7` but
* in the `neg` version the assertion is `False` as there is no `V ...` term (and so LH should say UNSAFE).

facundominguez commented:
I discovered some tests that worsen dramatically when inlining refinement types (i.e. transformation (1) in the description of this PR). Some later part of the pipeline is quite sensitive to the inlining of formulas.

`Tests.Prover.without_ple_pos.MapFusion.hs` consumes all the memory in my system and fails if I inline `lq_anf` bindings unrestrictedly. Not inlining `lq_anf`, however, causes `Tests.Benchmarks.bytestring.Data/ByteString/Fusion.T.hs` to improve less.

Both tests seem to do ok if I restrict inlining to refinement types with no more than two conjuncts. However, then `Tests.Prover.without_ple_pos.FoldrUniversal.hs` takes all of the memory in my system even inlining types with one conjunct only. 

I guess one other thing to try here is taking into account how many times a binding is used before inlining it, but I also wonder if there are other places where we can do less sophisticated optimizations and profit more.

My current plan is not to insert inlining in the verification pipeline, all the prunning of formulas will remain though. And I also will keep using inlining in the prettified output.

I'll continue fixing tests. I have figured out a bunch already. There remain only 4 failing tests, all of them due to the PIff simplification. Maybe I can get them all sorted next week.

facundominguez commented:
@ranjitjhala, if I rebase `develop`  here, I would lose the ability to run LH tests. Apparently LH needs to be updated after the move from `binary` to `store`.

ranjitjhala commented:
Oops did I not do that? Let me check!

On Fri, Aug 6, 2021 at 4:17 PM Facundo Domínguez ***@***.***>
wrote:

> @ranjitjhala
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ranjitjhala&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=P103Z_E7fVq25HVRaL3nLJ55p3j0EVYuFJA0qNkfrTM&s=11Tay36Chm1zL2IkeLZbXry3YPEuROFGKyUoYYzJetU&e=>,
> if I rebase develop here, I would lose the ability to run LH tests.
> Apparently LH needs to be updated after the move from binary to store.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquid-2Dfixpoint_pull_473-23issuecomment-2D894562384&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=P103Z_E7fVq25HVRaL3nLJ55p3j0EVYuFJA0qNkfrTM&s=ylZzh14RoS18cqwTqc4S_QXCE-HdPbtF820UzEivQt8&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OHKKWXD4PXMPSDH44DT3RUQLANCNFSM5AOMGCSA&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=P103Z_E7fVq25HVRaL3nLJ55p3j0EVYuFJA0qNkfrTM&s=r_dkc91u5J4emGDaOflmrZH9oIHf9P3ZFvQF7ENuri0&e=>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=P103Z_E7fVq25HVRaL3nLJ55p3j0EVYuFJA0qNkfrTM&s=XwFmAd3ak7R-H5a7BxHc2WnKLaNp1qYCRRPq5-ren7Q&e=>
> or Android
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26utm-5Fcampaign-3Dnotification-2Demail&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=P103Z_E7fVq25HVRaL3nLJ55p3j0EVYuFJA0qNkfrTM&s=aX2rYeSAYUGzpC2mHayNgngtmymNcYXHJ-dEM3kvJl4&e=>
> .
>

ranjitjhala commented:
@facundominguez -- I believe I merged in the edits to LH here 

https://github.com/ucsd-progsys/liquidhaskell/pull/1865

Maybe you have to rebase your LH too?

Or were you referring to something else?

facundominguez commented:
The last time I tried to build, I got some errors about missing `Binary` instances when building LH. I'll get back with the specifics if I can reproduce it.

facundominguez commented:
First issue when trying to rebase: `stack` refuses to build if liquid-fixpoint is updated to `liquid-fixpoint/develop`

```
$ git fetch origin -p
$ git checkout origin/develop
$ cd liquid-fixpoint
$ git fetch origin -p
$ git checkout origin/develop
$ cd ..
$ stack build liquid-platform
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for liquid-fixpoint-0.8.10.2.1:
    store needed, but the stack configuration has no specified version  (latest matching version is 0.7.11)
needed since liquid-fixpoint is a build target.

Some different approaches to resolving this:

  * Recommended action: try adding the following to your extra-deps in /home/facundo/tweag/liquidhaskell/stack.yaml:

- store-0.7.11@sha256:e12694225bc2165b31fc2bd3bb42777588c972e0d9bf58a20794c7a9324c6c4b,8071

Plan construction failed.
```

ranjitjhala commented:
How odd - wonder how this got past the CI?

ranjitjhala commented:
hmm strange, I had made all those edits, but oddly they were not on github -- not sure what's going on!

ranjitjhala commented:
I just pushed this

https://github.com/ucsd-progsys/liquidhaskell/commit/4500b6799eb7eb87cb344ecc2eed122297ed53b7 

hopefully I haven't broken everything - at least LH builds locally with `stack build`

facundominguez commented:
That worked for me. Thanks!

facundominguez commented:
I think this PR is as good as I can get it, at least without further feedback. All tests pass in LH, though the results are not so spectacular as we found initially.

The main differences between then and now are:
* now I'm retaining the bindings that can be used potentially by the kvars appearing in a constraint, which makes the environments a bit larger,
* now I don't inline lq_anf variables, which did speed up some of the benchmarks but has a catastrophic effect on [other tests](https://github.com/ucsd-progsys/liquid-fixpoint/pull/473#issuecomment-894515848).
* I moved from reducing `SInfo`s to reducing `FInfo`s. This helps avoiding issues with the ordering of lhs bindings (see haddock comments of the function `reduceEnvironments`).
* I created two passes over FInfos.  A first pass that merges and simplifies bindings defining the same symbol (function `simplifyBindings`), and a second pass to drop the irrelevant bindings (function `reduceEnvironments`). This made the passes easier to express than doing both in a single pass.

There are four tests which worsen between 1 and 2 seconds. These are tests where environment reduction can't reduce much because of the structure of dependencies between the constraints. Some of them got reductions from #480, though.

The rest of the tests improve or do not worsen more than a second. [Here](https://github.com/ucsd-progsys/liquid-fixpoint/files/6954585/summary_envred.ods) are the detailed comparisons, and here are the more pronounced differences:

test|develop|env-reduction|difference (seconds)
-- | -- | -- | --
Tests/Micro/ple-pos/STLC2.hs | 9.7845 | 11.4047 | 1.6202
Tests/Benchmarks/cse230/STLC.lhs | 9.6944 | 11.1946 | 1.5002
Tests/Benchmarks/esop/Base.hs | 53.5345 | 54.72 | 1.1855
Tests/Micro/ple-pos/STLC1.hs | 5.5741 | 6.6845 | 1.1104
Tests/Macro/unit-neg/StateConstraints0.hs | 18.7044 | 19.3544 | 0.65
Tests/Benchmarks/text/Data/Text/Fusion.hs | 16.1245 | 15.3246 | -0.7999
Tests/Benchmarks/bytestring/Data/ByteString/Internal.hs | 8.8447 | 7.9045 | -0.9401
Tests/Micro/ple-pos/RegexpDerivative.hs | 7.1244 | 6.0145 | -1.1099
Tests/Benchmarks/text/Data/Text/Lazy/Search.hs | 20.9445 | 18.7947 | -2.1498
Tests/Benchmarks/bytestring/Data/ByteString/Char8.hs | 12.8245 | 10.4445 | -2.38
Tests/Benchmarks/text/Data/Text/Encoding.hs | 61.8344 | 58.7645 | -3.0699
Tests/Benchmarks/text/Data/Text.hs | 28.5946 | 25.2847 | -3.3099
Tests/Prover/foundations/Lists.hs | 19.6443 | 16.2443 | -3.4
Tests/Benchmarks/bytestring/Data/ByteString/Lazy.hs | 37.4042 | 31.9544 | -5.4498
Tests/Benchmarks/text/Data/Text/Lazy/Encoding.hs | 46.5344 | 39.8643 | -6.6701
Tests/Benchmarks/bytestring/Data/ByteString/Fusion.T.hs | 42.6044 | 27.9745 | -14.6299


For all the code that I've introduced, the performance improvements are rather modest. The effort did help making the prettified output more accurate, and integrating it in the verification pipeline can help keeping the environments smaller and easier to debug. There's also the possibility that the performance balance evolves a bit over time as LH is tried on more code bases.

Other things that we can do in this area: reduce qualifiers, matches and define clauses and further improve the prettified output to make it easier to read (shorter variable names, include relevant defines's and matches).

With that said, let me thank you for your help this far while I acquaint with the code and the various tests. My understanding of `liquid-fixpoint` has quite evolved since I started this PR.

facundominguez commented:
Made some small fixes for the sake of correctness, though they don't affect tests or benchmarks significantly. I updated the numbers in the previous comment.

ranjitjhala commented:
Thanks @facundominguez !!! Bit unfortunate re: the effects, but no matter, this is still great.

Of "The main differences between then and now" are there some them more responsible for the _benefits_ and _slowdowns_ than the others? e.g. is the `lq_anf` inlining a single biggest factor? 

If so what do you think about making a flag for it (because the gains were _quite_ substantial I recall?)

Happy to have a zoom later this week if you think it would help? [Of course can also merge as is and restore those later if you think better?]

facundominguez commented:
> e.g. is the lq_anf inlining a single biggest factor?

I think this is true, together with the first item about retaining bindings needed by kvars.

> If so what do you think about making a flag for it (because the gains were quite substantial I recall?)

I'm a bit concerned that it can go so wrong. I'd rather work to make LH less prone to fail, but this flag is easy enough to put on the implementation if you want it.

> Of course can also merge as is and restore those later if you think better?

Merging this and dealing with it in a follow up PR makes the less maintenance work, I think. This PR does quite a bit already :)

ranjitjhala commented:
Great, lets merge this now and do a flag in a subsequent PR? I think it would be useful to have a flag to ultimately figure out how to make LH less prone to fail too...

facundominguez commented:
> Great, lets merge this now and do a flag in a subsequent PR?

Works for me.

ranjitjhala commented:
Just so I follow, 

* the retaining KVar binders affects _correctness_ (i.e. makes LH say unsafe when it should be safe) 
* the `lq_anf` inlining affects _performance_ (i.e. makes some tests blow up) ?

facundominguez commented:
Both statements are correct.
